File: backend/backend_JS/src/models/Product.ts
import mongoose, { Schema, Document } from 'mongoose';

interface IPrice {
  current: number;
  range: {
    min: number;
    max: number;
  };
}

interface IProductDetails {
  name: string;
  product: string;
  origin: string;
  productionCapacity: string;
  exportVolume: string;
  formAndCut: string;
  color: string;
  cultivationType: string;
  moisture?: string;
  forecast?: string;
}

interface IShippingDetails {
  hsCode: string;
  minQuantity: string;
  packaging: string;
  transportMode: string;
  incoterms: string;
  shelfLife: string;
}

export interface IProduct extends Document {
  productId: string;
  productName: string;
  images: string[];
  sellerId: string;
  price: IPrice;
  details: IProductDetails;
  shipping: IShippingDetails;
}

const PriceSchema: Schema = new Schema({
  current: { type: Number, required: true },
  range: {
    min: { type: Number, required: true },
    max: { type: Number, required: true }
  }
}, { _id: false });

const ProductDetailsSchema: Schema = new Schema({
  name: { type: String, required: true },
  product: { type: String, required: true },
  origin: { type: String, required: true },
  productionCapacity: { type: String },
  exportVolume: { type: String },
  formAndCut: { type: String },
  color: { type: String },
  cultivationType: { type: String },
  moisture: { type: String },
  forecast: { type: String }
}, { _id: false });

const ShippingDetailsSchema: Schema = new Schema({
  hsCode: { type: String },
  minQuantity: { type: String },
  packaging: { type: String },
  transportMode: { type: String },
  incoterms: { type: String },
  shelfLife: { type: String }
}, { _id: false });

const ProductSchema: Schema = new Schema({
  productId: { type: String, required: true, unique: true },
  productName: { type: String, required: true },
  images: [{ type: String }],
  sellerId: { type: String, required: true, index: true },
  price: { type: PriceSchema, required: true },
  details: { type: ProductDetailsSchema, required: true },
  shipping: { type: ShippingDetailsSchema, required: true }
}, { 
  timestamps: true,
  toJSON: { virtuals: false },
  toObject: { virtuals: false }
});

export default mongoose.model<IProduct>('Product', ProductSchema); 

File: backend/backend_JAVA/src/main/java/backend/models/Product.java
package backend.models;

import jakarta.validation.Valid;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import java.util.List;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "products")
public class Product {
    
    @Id
    private String id;
    
    @NotBlank(message = "ProductId is required")
    @Indexed(unique = true)
    private String productId;
    
    @NotBlank(message = "Product name is required")
    private String productName;
    
    @NotEmpty(message = "At least one product image is required")
    private List<String> images;
    
    @NotBlank(message = "SellerId is required")
    @Indexed
    private String sellerId;
    
    @NotNull(message = "Price information is required")
    @Valid
    private Price price;
    
    @NotNull(message = "Product details are required")
    @Valid
    private ProductDetails details;
    
    @NotNull(message = "Shipping details are required")
    @Valid
    private ShippingDetails shipping;
    
    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class Price {
        private double current;
        
        @NotNull(message = "Price range is required")
        @Valid
        private PriceRange range;
        
        @Data
        @NoArgsConstructor
        @AllArgsConstructor
        public static class PriceRange {
            private double min;
            private double max;
        }
    }
    
    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class ProductDetails {
        @NotBlank(message = "Product name is required")
        private String name;
        
        @NotBlank(message = "Product type is required")
        private String product;
        
        @NotBlank(message = "Origin country is required")
        private String origin;
        
        @NotBlank(message = "Production capacity is required")
        private String productionCapacity;
        
        @NotBlank(message = "Export volume is required")
        private String exportVolume;
        
        @NotBlank(message = "Form and cut information is required")
        private String formAndCut;
        
        @NotBlank(message = "Color information is required")
        private String color;
        
        @NotBlank(message = "Cultivation type is required")
        private String cultivationType;
        
        private String moisture;
        private String forecast;
    }
    
    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class ShippingDetails {
        @NotBlank(message = "HS code is required")
        private String hsCode;
        
        @NotBlank(message = "Minimum quantity is required")
        private String minQuantity;
        
        @NotBlank(message = "Packaging information is required")
        private String packaging;
        
        @NotBlank(message = "Transport mode is required")
        private String transportMode;
        
        @NotBlank(message = "Incoterms are required")
        private String incoterms;
        
        @NotBlank(message = "Shelf life is required")
        private String shelfLife;
    }
} 
